packages/website/ts/components/inputs/allowance_state_toggle.tsx

```diff
@@ -0,0 +1,158 @@
+import { utils } from 'ts/utils/utils';
+
+export interface AllowanceStateToggleProps {
+    networkId: number;
+    blockchain: Blockchain;
+    dispatcher: Dispatcher;
+    token: Token;
+    tokenState: TokenState;
+    userAddress: string;
+    onErrorOccurred?: (errType: BalanceErrs) => void;
+    refetchTokenStateAsync: () => Promise<void>;
+    tooltipDirection?: PointerDirection;
+}
+
+export interface AllowanceStateToggleState {
+    allowanceState: AllowanceState;
+    prevTokenState: TokenState;
+}
+
+const DEFAULT_ALLOWANCE_AMOUNT_IN_BASE_UNITS = new BigNumber(2).pow(256).minus(1);
+
+export class AllowanceStateToggle extends React.Component<AllowanceStateToggleProps, AllowanceStateToggleState> {
+    public static defaultProps = {
+        onErrorOccurred: _.noop.bind(_),
+        tooltipDirection: PointerDirection.Right,
+    };
+    private static _getAllowanceState(tokenState: TokenState): AllowanceState {
+        if (!tokenState.isLoaded) {
+            return AllowanceState.Loading;
+        }
+        if (tokenState.allowance.gt(0)) {
+            return AllowanceState.Unlocked;
+        }
+        return AllowanceState.Locked;
+    }
+    constructor(props: AllowanceStateToggleProps) {
+        super(props);
+        const tokenState = props.tokenState;
+        this.state = {
+            allowanceState: AllowanceStateToggle._getAllowanceState(tokenState),
+            prevTokenState: tokenState,
+        };
+    }
+
+    public render(): React.ReactNode {
+        const tooltipId = `tooltip-id-${this.props.token.symbol}`;
+        return (
+            <Container cursor="pointer">
+                <ReactTooltip id={tooltipId} effect="solid" offset={{ top: 3 }}>
+                    {this._getTooltipContent()}
+                </ReactTooltip>
+                <div
+                    data-tip={true}
+                    data-for={tooltipId}
+                    data-place={this.props.tooltipDirection}
+                    onClick={this._onToggleAllowanceAsync.bind(this)}
+                >
+                    <AllowanceStateView allowanceState={this.state.allowanceState} />
+                </div>
+            </Container>
+        );
+    }
+    public componentWillReceiveProps(nextProps: AllowanceStateToggleProps): void {
+        const nextTokenState = nextProps.tokenState;
+        const prevTokenState = this.state.prevTokenState;
+        if (
+            !nextTokenState.allowance.eq(prevTokenState.allowance) ||
+            nextTokenState.isLoaded !== prevTokenState.isLoaded
+        ) {
+            const tokenState = nextProps.tokenState;
+            this.setState({
+                prevTokenState: tokenState,
+                allowanceState: AllowanceStateToggle._getAllowanceState(nextTokenState),
+            });
+        }
+    }
+    private _getTooltipContent(): React.ReactNode {
+        const symbol = this.props.token.symbol;
+        const isLockingToken = this.props.tokenState.allowance.gt(0);
```

Comment:
can we put this in the Loading case or AND it with `this.state.allowanceState === AllowanceState.Loading`

Reply:
I think we need to handle the case when the tokenState is loading its initial state, when we are doing the initial load the tooltip incorrectly displays "locking/unlocking X" when really we just are waiting to read the state from the blockchain